contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -0,0 +1,297 @@
+     * @param conversionFactor 1e18 if seeking the ETH price, and a 6 decimal ETH-USDC price in the case of other assets
+     */
+    function fetchAnchorPrice(TokenConfig memory config, uint conversionFactor) internal virtual returns (uint) {
+        (uint nowCumulativePrice, uint oldCumulativePrice, uint oldTimestamp) = pokeWindowValues(config);
+
+        // This should be impossible, but better safe than sorry
+        require(block.timestamp > oldTimestamp, "now must come after before");
+        uint timeElapsed = block.timestamp - oldTimestamp;
+
+        // Calculate uniswap time-weighted average price
+        FixedPoint.uq112x112 memory priceAverage = FixedPoint.uq112x112(uint224((nowCumulativePrice - oldCumulativePrice) / timeElapsed));
+        uint rawUniswapPriceMantissa = mul(priceAverage.decode112with18(), conversionFactor);
+        uint anchorPrice;
+
+        // Adjust rawUniswapPrice according to the units of the non-ETH asset
+        // In the case of ETH, we would have to scale by 1e6 / USDC_UNITS, but since baseUnit2 is 1e6 (USDC), it cancels
+        if (config.isUniswapReversed) {
+            // rawUniswapPriceMantissa * ethBaseUnit / config.baseUnit / expScale, but we simplify bc ethBaseUnit == expScale
+            anchorPrice = rawUniswapPriceMantissa / config.baseUnit;
+        } else {
+            anchorPrice = mul(rawUniswapPriceMantissa, config.baseUnit) / ethBaseUnit / expScale;
+        }
+
+        emit AnchorPriceUpdate(config.uniswapMarket, anchorPrice, nowCumulativePrice, oldCumulativePrice, oldTimestamp);
+
+        return anchorPrice;
+    }
+
+    /**
+     * @dev Get time-weighted average prices for a token at the current timestamp.
+     *  Update new and old observations of lagging window if period elapsed.
+     */
+    function pokeWindowValues(TokenConfig memory config) internal returns (uint, uint, uint) {
+        address uniswapMarket = config.uniswapMarket;
+        uint cumulativePrice = currentCumulativePrice(config);
+
+        Observation memory newObservation = newObservations[uniswapMarket];
+        Observation memory oldObservation = oldObservations[uniswapMarket];
+
+        // Update new and old observations if elapsed time is greater than or equal to anchor period
+        uint timeElapsed = block.timestamp - newObservation.timestamp;
+        if (timeElapsed >= anchorPeriod) {
+            emit UniswapWindowUpdate(uniswapMarket, oldObservation.timestamp, newObservation.timestamp, oldObservation.acc, newObservation.acc);
+            oldObservations[uniswapMarket].timestamp = newObservation.timestamp;
+            oldObservations[uniswapMarket].acc = newObservation.acc;
+
+            newObservations[uniswapMarket].timestamp = block.timestamp;
+            newObservations[uniswapMarket].acc = cumulativePrice;
+        }
+        return (cumulativePrice, oldObservation.acc, oldObservation.timestamp);
```

Comment:
:corn: caught a regression here (fixing), this returns the old `oldObservation` now, not the new one